### PR TITLE
Fix bad self casts when calling methods of final types

### LINKED ISSUE #2747: 'final' subclasses produce suboptimal self casts
Starting from:
```cython
# subclass.pyx
cdef class A:
    cdef int foo(self):
        return -1

    cdef int bar(self, int x):
        return -1

cdef class Sub(A):
    cdef int foo(self):
        return self.bar(0)

    cdef int bar(self, int x):
        return x + 1
```
compiles fine with `cythonize -b subclass.pyx`. If you add `@cython.final` onto `Sub`, then it produces a non-fatal warning:
```
gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -m64 -fPIC -m64 -fPIC -fPIC -Iconda/envs/cp36/include/python3.6m -c subclass.c -o build/temp.linux-x86_64-3.6/subclass.o
subclass.c: In function ‘__pyx_f_8subclass_3Sub_foo’:
subclass.c:1667:41: warning: passing argument 1 of ‘__pyx_f_8subclass_3Sub_bar’ from incompatible pointer type [-Wincompatible-pointer-types]
   __pyx_r = __pyx_f_8subclass_3Sub_bar(((struct __pyx_obj_8subclass_A *)__pyx_v_self), 0);
                                        ~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
subclass.c:1205:85: note: expected ‘struct __pyx_obj_8subclass_Sub *’ but argument is of type ‘struct __pyx_obj_8subclass_A *’
 static int __pyx_f_8subclass_3Sub_bar(CYTHON_UNUSED struct __pyx_obj_8subclass_Sub *__pyx_v_self, int __pyx_v_x); /* proto*/
                                                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~
```
But if you _also_ add `# distutils: language = c++` to the file to compile as C++, then it fails to compile:
```
gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -m64 -fPIC -m64 -fPIC -fPIC -Iconda/envs/cp36/include/python3.6m -c subclass.cpp -o build/temp.linux-x86_64-3.6/subclass.o
subclass.cpp: In function ‘int __pyx_f_8subclass_3Sub_foo(__pyx_obj_8subclass_Sub*)’:
subclass.cpp:1682:41: error: cannot convert ‘__pyx_obj_8subclass_A*’ to ‘__pyx_obj_8subclass_Sub*’
   __pyx_r = __pyx_f_8subclass_3Sub_bar(((struct __pyx_obj_8subclass_A *)__pyx_v_self), 0);
                                        ~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
subclass.cpp:1220:85: note:   initializing argument 1 of ‘int __pyx_f_8subclass_3Sub_bar(__pyx_obj_8subclass_Sub*, int)’
 static int __pyx_f_8subclass_3Sub_bar(CYTHON_UNUSED struct __pyx_obj_8subclass_Sub *__pyx_v_self, int __pyx_v_x); /* proto*/
                                                                                     ^
error: command 'gcc' failed with exit status 1
```

### PR DESCRIPTION
**Pull request by dalcinl**
May not be the perfect and definitive fix, but hopefully better than adding workarounds with `#pragma GCC ...`.

Fixes #2747.

### CODE REVIEW

File: Cython/Compiler/Symtab.py [hunk 1]
```diff
@@ -2517,8 +2517,11 @@ def declare_cfunction(self, name, type, pos,
                 if entry.is_final_cmethod and entry.is_inherited:
                     error(pos, "Overriding final methods is not allowed")
                 elif type.same_c_signature_as(entry.type, as_cmethod = 1) and type.nogil == entry.type.nogil:
-                    # Fix with_gil vs nogil.
-                    entry.type = entry.type.with_with_gil(type.with_gil)
+                    if self.parent_type.is_final_type:
+                        entry = self.add_cfunction(name, type, pos, cname, visibility='ignore', modifiers=modifiers)
+                    else:
+                        # Fix with_gil vs nogil.
+                        entry.type = entry.type.with_with_gil(type.with_gil)
```
Comment (da-woods) [inline Cython/Compiler/Symtab.py:2524]: I just wonder if something to test this should be included in the test? It isn't obvious to me whether this should be inside an `else` or not.

Comment (dalcinl) [reply Cython/Compiler/Symtab.py:2524]: This whole thing confuses me as well. I have not idea what to exactly test for. I just assumed that if a new final method is being added to a scopy, the input type information and its gil properties are already properly set, therefore you just create a new entry. 

The else branch is just leaving things as they were before, without messing with them in any way. I'm just trying to fix the underlying issue the right way, not with a low-quality temporary workaround as before, while minimizing the risk of breaking something. It looks really weird that this else branch would mess in such a way with `entry.type` in for an already-declared entry that comes from the parent type scope. But I'm admittedly not an expert, therefore I'm not going to mess with what was already there.

**Comment by dalcinl** [comment]
@da-woods I resolved your previous comment after a force-push. 

What I'm doing now seems to be much cleaner. This indeed seems to be fixing a long-standing issue with the implementation of `@cython.final`, that is, the concrete type of a final method was not being properly set in the entry. 

* I'm not 100% sure I should treat inline methods the same way as final, but the codegen consequences seem to be quite similar . 
* Adding `vtable_type` to the entry looks a bit hackish, is there a better way/place that you can think of?

Your feedback is most needed...

File: Cython/Compiler/Symtab.py [hunk 2]
```diff
@@ -2551,6 +2551,9 @@ def declare_cfunction(self, name, type, pos,
         if self.parent_type.is_final_type or entry.is_inline_cmethod or self.directives.get('final'):
             entry.is_final_cmethod = True
             entry.final_func_cname = entry.func_cname
+            if not type.is_fused:
```
Comment (da-woods) [inline Cython/Compiler/Symtab.py:2554]: I'm a little suspicious of excluding fused types, but I think there are probably other issues with fused methods (https://github.com/cython/cython/pull/5931) so maybe we should worry too much right now and sort it out later

**Review by da-woods — COMMENTED** [review]
Couple of comments but looks basically good to me, and seems to generate the right code.

File: Cython/Compiler/Symtab.py (hunk 2 above)
Comment (dalcinl) [reply Cython/Compiler/Symtab.py:2554]: I had to exclude them here because otherwise at C code generation time, the fused type is asked for the C signature, which leads to a explicit compiler failure. 
I believe addressing fused methods in full should be for another rodeo, indeed.

File: Cython/Compiler/ModuleNode.py [hunk 3]
```diff
@@ -3931,7 +3931,8 @@ def generate_exttype_vtable_init_code(self, entry, code):
                 if entry.func_cname]
             if c_method_entries:
                 for meth_entry in c_method_entries:
-                    cast = meth_entry.type.signature_cast_string()
+                    vtable_type = getattr(meth_entry, 'vtable_type', meth_entry.type)
```
Comment (scoder) [inline Cython/Compiler/ModuleNode.py:3934]: What we usually do instead of a `getattr()` is to add the new attribute normally as a class (default) attribute in `Entry` and then set it only when we have something reasonable to assign it to. That way, it always exists.

File: test-requirements-312.txt [hunk 4]
```diff
@@ -1,4 +1,4 @@
-numpy
+numpy<2
```
Comment (scoder) [inline test-requirements-312.txt:1]: This seems unrelated. (Same below.)

**Review by scoder — CHANGES_REQUESTED** [review]

**Review by scoder — APPROVED** [review]
LGTM, thanks.

**Comment by dalcinl** [comment]
@scoder Sorry, but I don't know the current dynamics. Are you going to merge this PR directly to 3.0.x clicking the button here in the GitHub UI, then you manually merge the branch back into master and push using the cli locally? Or should I submit the same fix in a new PR against master?

**Comment by scoder** [comment]
We still merge 3.0.x upwards, so merging this PR is fine.

**Comment by hroncok** [comment]
For the record, I see:

    src/gevent/queue.c:11895:114: error: passing argument 4 of ‘__pyx_vtabptr_6gevent_14_gevent_cqueue_UnboundQueue->__pyx_base.put’ from incompatible pointer type [-Wincompatible-pointer-types]

In gevent with Cython 3.0.10 (not present before).

https://github.com/gevent/gevent/issues/2031

**Comment by dalcinl** [comment]
It took me an incredible amount of effort to come up with a trivial reproducer

```cython
# queue.pxd
cimport cython

cdef class Queue:
    cpdef put(self, item=*)

@cython.final
cdef class UnboundQueue(Queue):
    pass
```

```python
# queue.py
class Queue(object):
    def __init__(self): pass
    def put(self, item=None): pass

class UnboundQueue(Queue):
    def __init__(self): pass
    def put(self, item=None): pass
```

Slight changes to the code (e.g. remove the `__init__` method in `UnboundQueue`, or not use a default arg in `put()`) no longer reproduce the problem. This all sounds like this is hitting quite of a corner case. I need more time to investigate.

PS: An obvious quick "fix" for gevent is to not use `@cython.final` in the `UnboundQueue` class.

**Comment by stratakis** [comment]
gevent indeed builds fine with removing the `@cython.final` in the `UnboundQueue`.

**Comment by dalcinl** [comment]
I've not been able to figure out what's going on :disappointed:
